### ISSUE TITLE
E-Mail Challenge/Response broken with Python 3.8 (python-twisted 19.7.0-4)

### ISSUE THREAD

**Issue by Obrekr** [matrix-org/synapse#6393]
### Description

Matrix Server running with python 3.8 fails to send mail for 3pid mail verification.

### Steps to reproduce

- Update to python 3.8
- Register a new user with 3pid e-mail required on homeserver with a Matrix client

Here is the log:
```
Nov 20 16:44:26 matrix synapse[1393]: 2019-11-20 16:44:26,313 - synapse.access.http.8008 - 228 - INFO - POST-35893 - 37.24.110.62 - 8008 - Received request: POST /_matrix/client/r0/register/email/requestToken
Nov 20 16:44:26 matrix synapse[1393]: 2019-11-20 16:44:26,327 - synapse.push.mailer - 314 - INFO - POST-35893 - Sending email to <deleted>
Nov 20 16:44:26 matrix synapse[1393]: 2019-11-20 16:44:26,382 - twisted - 172 - CRITICAL -  - Unhandled Error
Nov 20 16:44:26 matrix synapse[1393]: Traceback (most recent call last):
Nov 20 16:44:26 matrix synapse[1393]:   File "/usr/lib/python3.8/site-packages/twisted/python/log.py", line 103, in callWithLogger
Nov 20 16:44:26 matrix synapse[1393]:     return callWithContext({"system": lp}, func, *args, **kw)
Nov 20 16:44:26 matrix synapse[1393]:   File "/usr/lib/python3.8/site-packages/twisted/python/log.py", line 86, in callWithContext
Nov 20 16:44:26 matrix synapse[1393]:     return context.call({ILogContext: newCtx}, func, *args, **kw)
Nov 20 16:44:26 matrix synapse[1393]:   File "/usr/lib/python3.8/site-packages/twisted/python/context.py", line 122, in callWithContext
Nov 20 16:44:26 matrix synapse[1393]:     return self.currentContext().callWithContext(ctx, func, *args, **kw)
Nov 20 16:44:26 matrix synapse[1393]:   File "/usr/lib/python3.8/site-packages/twisted/python/context.py", line 85, in callWithContext
Nov 20 16:44:26 matrix synapse[1393]:     return func(*args,**kw)
Nov 20 16:44:26 matrix synapse[1393]: --- <exception caught here> ---
Nov 20 16:44:26 matrix synapse[1393]:   File "/usr/lib/python3.8/site-packages/twisted/internet/posixbase.py", line 614, in _doReadOrWrite
Nov 20 16:44:26 matrix synapse[1393]:     why = selectable.doRead()
Nov 20 16:44:26 matrix synapse[1393]:   File "/usr/lib/python3.8/site-packages/twisted/internet/tcp.py", line 243, in doRead
Nov 20 16:44:26 matrix synapse[1393]:     return self._dataReceived(data)
Nov 20 16:44:26 matrix synapse[1393]:   File "/usr/lib/python3.8/site-packages/twisted/internet/tcp.py", line 249, in _dataReceived
Nov 20 16:44:26 matrix synapse[1393]:     rval = self.protocol.dataReceived(data)
Nov 20 16:44:26 matrix synapse[1393]:   File "/usr/lib/python3.8/site-packages/twisted/protocols/tls.py", line 330, in dataReceived
Nov 20 16:44:26 matrix synapse[1393]:     self._flushReceiveBIO()
Nov 20 16:44:26 matrix synapse[1393]:   File "/usr/lib/python3.8/site-packages/twisted/protocols/tls.py", line 295, in _flushReceiveBIO
Nov 20 16:44:26 matrix synapse[1393]:     ProtocolWrapper.dataReceived(self, bytes)
Nov 20 16:44:26 matrix synapse[1393]:   File "/usr/lib/python3.8/site-packages/twisted/protocols/policies.py", line 120, in dataReceived
Nov 20 16:44:26 matrix synapse[1393]:     self.wrappedProtocol.dataReceived(data)
Nov 20 16:44:26 matrix synapse[1393]:   File "/usr/lib/python3.8/site-packages/twisted/protocols/basic.py", line 572, in dataReceived
Nov 20 16:44:26 matrix synapse[1393]:     why = self.lineReceived(line)
Nov 20 16:44:26 matrix synapse[1393]:   File "/usr/lib/python3.8/site-packages/twisted/mail/smtp.py", line 996, in lineReceived
Nov 20 16:44:26 matrix synapse[1393]:     why = self._okresponse(self.code, b'\n'.join(self.resp))
Nov 20 16:44:26 matrix synapse[1393]:   File "/usr/lib/python3.8/site-packages/twisted/mail/smtp.py", line 1557, in esmtpState_challenge
Nov 20 16:44:26 matrix synapse[1393]:     self._authResponse(self._authinfo, resp)
Nov 20 16:44:26 matrix synapse[1393]:   File "/usr/lib/python3.8/site-packages/twisted/mail/smtp.py", line 1570, in _authResponse
Nov 20 16:44:26 matrix synapse[1393]:     resp = auth.challengeResponse(self.secret, challenge)
Nov 20 16:44:26 matrix synapse[1393]:   File "/usr/lib/python3.8/site-packages/twisted/mail/_cred.py", line 31, in challengeResponse
Nov 20 16:44:26 matrix synapse[1393]:     response = hmac.HMAC(secret, chal).hexdigest().encode('ascii')
Nov 20 16:44:26 matrix synapse[1393]:   File "/usr/lib/python3.8/hmac.py", line 49, in __init__
Nov 20 16:44:26 matrix synapse[1393]:     raise ValueError('`digestmod` is required.')
Nov 20 16:44:26 matrix synapse[1393]: builtins.ValueError: `digestmod` is required.
Nov 20 16:44:26 matrix synapse[1393]: 2019-11-20 16:44:26,388 - twisted - 172 - INFO -  - SMTP Client retrying server. Retry: 5
```

[Python Doc 3.7](https://docs.python.org/3.7/library/hmac.html) states that default digest for hmac will be deprecated as of python 3.8.
<!--
Describe how what happens differs from what you expected.

If you can identify any relevant log snippets from _homeserver.log_, please include
those (please be careful to remove any personal or private data). Please surround them with
``` (three backticks, on a line on their own), so that they are formatted legibly.
-->

### Version information

<!-- IMPORTANT: please answer the following questions, to help us narrow down the problem -->

<!-- Was this issue identified on matrix.org or another homeserver? -->
- **Homeserver**:
Own homeserver running with 3pid e-mail required.

<!--
 What version of Synapse is running?

You can find the Synapse version with this command:

$ curl http://localhost:8008/_synapse/admin/v1/server_version

(You may need to replace `localhost:8008` if Synapse is not configured to
listen on that port.)
-->
- **Version**:
```
$curl http://localhost:8008/_synapse/admin/v1/server_version
{
    "python_version": "3.8.0",
    "server_version": "1.5.1"
}
```

 ```
$pacman -Qs python-twisted                                  
local/python-twisted 19.7.0-4
    Asynchronous networking framework written in Python
```

```
$python --version
Python 3.8.0
```
- **Install method**:
<!-- examples: package manager/git clone/pip  -->
Arch Linux paket manager `pacman -S matrix-synapse`
- **Platform**:
Up-to-date Arch Linux.
<!--
Tell us about the environment in which your homeserver is operating
distro, hardware, if it's running in a vm/container, etc.
-->

**Comment by hawkowl** [comment]
Hi, thanks for the report.

This is a problem in upstream Twisted, which I've filed at https://twistedmatrix.com/trac/ticket/9733 . As Synapse and Twisted don't yet support Python 3.8, you'll have to use 3.5-3.7 in the meantime.

I'm going to close this issue in preference of the Twisted one, which is the root cause.

**Comment by mithodin** [comment]
I've created a pull request https://github.com/twisted/twisted/pull/1219 that fixes this. E-Mail challenge works with the patch applied, at least on my synapse setup.

**Comment by richvdh** [comment]
I think we should keep this open to track the issue as it applies to synapse, not least because Synapse now claims to support 3.8.

**Comment by mithodin** [comment]
My patch has been merged upstream, so this should be fixed soon.

**Comment by richvdh** [comment]
hopefully, this is fixed as of twisted 20.3.
